language_models: Add support for API key to Ollama provider

imumesh18 opened the pull request:
Closes https://github.com/zed-industries/zed/issues/19491

Release Notes:

- Add support for API key and API URL to connect to remote Ollama provider. This also adds URL and API key to be set from agent panel settings.

imumesh18 committed:
language_models: Add support for API key and API URL to ollama provider

imumesh18 committed:
Redesign Ollama configuration UI for improved clarity

imumesh18 committed:
fix clippy

m-saghir commented:
@ConradIrwin & @notpeter, I noticed that you had reviewed #19493, which addressed the same issue (#19491). I was wondering if this solution be appropriate? Thanks!

notpeter commented:
Any suggestions for how to test this? 
Are there any commercial providers I could test against?
Or does anyone have an example http proxy/config (nginx, caddy, whatever) to put in front of ollama?

notpeter committed:
Merge branch 'main' into push-trzsxukkukpr

notpeter committed:
Merge fix

imumesh18 commented:
Hey @notpeter here you go.

Setup guide for testing authentication for Ollama with [OpenWebUI](https://docs.openwebui.com/)
  - Run the following command to start OpenWebUI and connect it to your local Ollama instance:
    `docker run -d -p 3000:8080 --add-host=host.docker.internal:host-gateway -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:main`. This assumes your ollama server is running on host machine.
  - Go to `127.0.0.1:3000` in your browser and create an admin account.
  - In OpenWebUI, navigate to `Settings > Account` and copy your API key.
  - Set the URL to `http://localhost:3000/ollama` and use your copied API key.
  - For more information, see the Ollama Proxy documentation: https://docs.openwebui.com/getting-started/api-endpoints#-ollama-api-proxy-support

notpeter committed:
Merge branch 'main' into push-trzsxukkukpr

imumesh18 committed:
Merge remote-tracking branch 'origin/main' into push-trzsxukkukpr

imumesh18 commented:
@notpeter just in case you need another approach ollama just launched turbo mode in their new release: https://github.com/ollama/ollama/blob/main/docs/turbo.md. You can consider it for testing.

notpeter committed:
docs, settings text

notpeter committed:
Merge branch 'main' into push-trzsxukkukpr

notpeter commented:
Hi @imumesh18 I finally got a chance to check this out this morning.

1. I aded some basic docs.
2. I modified settings panel to make it little simpler and to correct the styling of the "reset" buttons.

I tested it with Ollama Turbo and got some tokens back but then it errored with
<img width="483" height="210" alt="Screenshot 2025-08-06 at 9 46 56" src="https://github.com/user-attachments/assets/318e5e85-00cb-4458-9336-f17b66c08438" />

I would love to pair on fixing this.
Feel free to reach out discord (@notpeter) if you're interested.

notpeter committed:
Two space gang.

imumesh18 commented:
Pinged you on discord @notpeter. Also I believe this could be a malformed response from ollama. Saw couple of them on launch day in their github issues. But happy to pair and fix in case something we are doing wrong from zed.

oliverbarnes commented:
Hey @imumesh18, I had been implementing api key support for Ollama edit predictions (but not for the agent nor the inline assistant) in parallel, on #33616, until @ghost91- pointed me to your PR. Looks like there's some overlap :)

I'd be happy to scrap my side of the overlap and just merge in your dedicated solution. Are you still struggling with the malformed response issue?

Thanks for the local testing steps, btw. I was in the process of setting up a Hostinger account just for this.

imumesh18 commented:
Hey @oliverbarnes, In local setup I haven't seen the issue of request parsing. It mostly occurs in ollama turbo(their hosted service). Unfortunately, I currently don't have the subscription for that and need to pair with peter to see if we can fix it. Afaik, that error should have nothing to do with auth but haven't debugged it yet so can't say is surely. For edit prediction I haven't gone through your latest changes and not aware of the implementation. But I am more than happy in case you want to make some changes to this PR to accommodate edit_prediction as well.

imumesh18 committed:
Merge remote-tracking branch 'origin/main' into push-trzsxukkukpr

imumesh18 committed:
📎

Signed-off-by: Umesh Yadav <git@umesh.dev>

oliverbarnes commented:
Thanks, I'm trying to think up what that accommodation would look like...

One key difference between the two implementations is how credentials are passed around. In the edit predictions PR both the api url and key are kept in a new `ollama::ollama_completion_provider::OllamaService`, while here they're in `language_models::provider::ollama::State`.

Could the new ollama edit prediction provider use `language_models::provider::ollama::State` instead?

To be honest I don't fully understand yet the separation between provider crates like `ollama` or `anthropic` and provider files under `language_models`. I'm guessing it's for historical reasons? As far as I tell, `language_models` holds agent and inline assistant functionality.

I implemented the ollama completion (edit prediction) provider under the `ollama` crate following the example of the`copilot` and `supermaven` implementation, effectively separating edit prediction functionality from agent and inline assistant in different crates.

I'm not sure whether that separation should be strictly enforced regarding api key management.

imumesh18 committed:
Move OLLAMA_API_KEY_VAR to ollama crate

Co-authored-by: Oliver Azevedo Barnes <oliver@liquidvoting.io>

imumesh18 committed:
Merge remote-tracking branch 'origin/main' into push-trzsxukkukpr

imumesh18 committed:
Merge branch 'main' into push-trzsxukkukpr

imumesh18 committed:
Merge branch 'main' into push-trzsxukkukpr

mgsloan committed:
Use api key placeholder text more like prefix of ollama turbo key

mgsloan committed:
Misc cleanup + no need to handle empty api_url (not done in other providers)

mgsloan committed:
Refactoring: pull out render helper methods

mgsloan committed:
Remove a redundant v_flex and fix vertical spacing

mgsloan committed:
Remove more redundant v_flex / div

mgsloan committed:
Remove "API URL configured. " prefix from set URL

mgsloan committed:
Swap order of url editor and API key editor

mgsloan committed:
Small label instead of XSmall for text mentioning env var, matching Mistral

mgsloan committed:
Remove loading_models_task - I don't see much benefit of this

mgsloan committed:
Clear available_models when URL changes, so it no longer appears "Connected"

mgsloan committed:
Polish instructions and always show them

imumesh18 committed:
Replace Indicator dot with check icon for success state

imumesh18 committed:
more fixes

Signed-off-by: Umesh Yadav <git@umesh.dev>

imumesh18 committed:
Fix Ollama authentication check logic

The auth check should only depend on available models since Ollama ollama does
fetch models after auth for remote and direct model fetch for local model.

Signed-off-by: Umesh Yadav <git@umesh.dev>

imumesh18 committed:
Merge branch 'main' into push-trzsxukkukpr

imumesh18 commented on docs/src/ai/llm-providers.md:
```diff
@@ -376,6 +376,20 @@ If the model is tagged with `thinking` in the Ollama catalog, set this option an
 
+#### Ollama Authentication
+
+In addition to running Ollama on your own hardware, which generally does not require authentication, Zed also supports connecting to Ollama API Keys are required for authentication.
```
Ohh nice catch. Thanks

imumesh18 committed:
Fix Ollama authentication section for clarity and accuracy

Signed-off-by: Umesh Yadav <git@umesh.dev>

mgsloan committed:
Merge branch 'main' into push-trzsxukkukpr

mgsloan committed:
Use new ApiKeyState helper + other refactors from #38163

mgsloan committed:
When url is reset, set it to the default instead of null in json

mgsloan committed:
Merge branch 'main' into push-trzsxukkukpr

mgsloan committed:
Use new SingleLineInput::set_text method

mgsloan committed:
Rename `State::available_models` to `State::fetched_models`

mgsloan committed:
Remove fetch of models on init - should be unnecessary

mgsloan committed:
Clear fetched_models and refetch when url or api key change

Causes "Connected" state to actually update

mgsloan commented:
Made a few more changes and now happy with the configuration UI behavior :)  Thanks a bunch for all your work on this @imumesh18 !
